src/accelerate/utils/modeling.py

```diff
@@ -698,7 +699,16 @@ def compute_module_sizes(
         special_dtypes = {key: _get_proper_dtype(dtyp) for key, dtyp in special_dtypes.items()}
         special_dtypes_size = {key: dtype_byte_size(dtyp) for key, dtyp in special_dtypes.items()}
     module_sizes = defaultdict(int)
-    for name, tensor in named_module_tensors(model, recurse=True):
+
+    module_list = []
+    
+    if not buffers_only:
+        module_list = named_module_tensors(model, recurse=True) 
+    else:
+        if hasattr(model, "named_buffers"): 
```

Comment:
But there is a test that won't not pass without this. I'm not sure if I should touch the code of that test.

Reply:
Could you link the test ?

Reply:
```
FAILED tests/test_modeling_utils.py::ModelingUtilsTester::test_infer_auto_device_map_with_tied_weights - AttributeError: 'Parameter' object has no attribute 'named_buffers'
``` 

detail log:

```
_________________________ ModelingUtilsTester.test_infer_auto_device_map_with_tied_weights _________________________

self = <test_modeling_utils.ModelingUtilsTester testMethod=test_infer_auto_device_map_with_tied_weights>

    def test_infer_auto_device_map_with_tied_weights(self):
        model = nn.Sequential(
            OrderedDict([("layer1", ModelForTest()), ("layer2", ModelForTest()), ("layer3", ModelForTest())])
        )
        model.layer3.linear2.weight = model.layer1.linear2.weight
        device_map = infer_auto_device_map(model, max_memory={0: 400, 1: 500})
        expected = {"layer1": 0, "layer3.linear2": 0, "layer2": 1, "layer3.linear1": 1, "layer3.batchnorm": 1}
        assert device_map == expected
    
        # With three weights tied together
        model.layer2.linear2.weight = model.layer1.linear2.weight
        device_map = infer_auto_device_map(model, max_memory={0: 400, 1: 500})
        expected = {
            "layer1": 0,
            "layer2.linear2": 0,
            "layer3.linear2": 0,
            "layer2.linear1": 1,
            "layer2.batchnorm": 1,
            "layer3.linear1": 1,
            "layer3.batchnorm": 1,
        }
        assert device_map == expected
    
        # With two groups of weights tied together
        model.layer2.linear1.weight = model.layer1.linear1.weight
        device_map = infer_auto_device_map(model, max_memory={0: 400, 1: 500})
        expected = {
            "layer1": 0,
            "layer2.linear1": 0,
            "layer2.linear2": 0,
            "layer3.linear2": 0,
            "layer2.batchnorm": 1,
            "layer3.linear1": 1,
            "layer3.batchnorm": 1,
        }
        assert device_map == expected
    
        # With weights ties in the same module
        model = nn.Sequential(
            OrderedDict(
                [
                    ("linear1", nn.Linear(4, 4)),
                    ("linear2", nn.Linear(6, 6)),
                    ("linear3", nn.Linear(4, 4)),
                    ("linear4", nn.Linear(6, 6)),
                ]
            )
        )
        model.linear3.weight = model.linear1.weight
        model.linear3.bias = model.linear1.bias
        device_map = infer_auto_device_map(model, max_memory={0: 250, 1: 400})
        expected = {"linear1": 0, "linear2": 1, "linear3": 0, "linear4": 1}
        assert device_map == expected
    
        # With tied weights sharing a same prefix name (`compute.weight` vs `compute.weight_submodule.parameter`)
        class SubModule(torch.nn.Module):
            def __init__(self, ref_to_parameter):
                super().__init__()
                self.parameter = ref_to_parameter
    
            def forward(self, x):
                return self.x + torch.max(self.parameter)
    
        class LinearModuleAndSubModule(torch.nn.Linear):
            def __init__(self, in_features, out_features):
                super().__init__(in_features, out_features)
                self.weight_submodule = SubModule(self.weight)
    
            def forward(self, x):
                return torch.nn.functional.linear(self.weight_submodule(x), self.weight)
    
        class Model(torch.nn.Module):
            def __init__(self):
                super().__init__()
                self.compute = LinearModuleAndSubModule(3, 8)
    
            def forward(self, x):
                return self.compute(x)
    
        model = Model()
    
        device_memory = {0: 4, "cpu": 96000}  # Low memory device, just to force splitting and trigger the error
>       infer_auto_device_map(model, device_memory)

tests/test_modeling_utils.py:581: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
src/accelerate/utils/modeling.py:1251: in infer_auto_device_map
    current_buffer_size = compute_module_total_buffer_size(
src/accelerate/utils/modeling.py:740: in compute_module_total_buffer_size
    module_sizes = compute_module_sizes(model, dtype, special_dtypes, True)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

model = Parameter containing:
tensor([[-0.1732, -0.2755,  0.5513],
        [ 0.3202, -0.3520, -0.2377],
        [ 0.1762,  0.4...0662, -0.2712, -0.3916],
        [-0.3685,  0.5621,  0.1996],
        [ 0.2687,  0.0723, -0.5561]], requires_grad=True)
dtype = None, special_dtypes = None, buffers_only = True

    def compute_module_sizes(
        model: nn.Module,
        dtype: Optional[Union[str, torch.device]] = None,
        special_dtypes: Optional[Dict[str, Union[str, torch.device]]] = None,
        buffers_only: bool = False,
    ):
        """
        Compute the size of each submodule of a given model.
        """
        if dtype is not None:
            dtype = _get_proper_dtype(dtype)
            dtype_size = dtype_byte_size(dtype)
        if special_dtypes is not None:
            special_dtypes = {key: _get_proper_dtype(dtyp) for key, dtyp in special_dtypes.items()}
            special_dtypes_size = {key: dtype_byte_size(dtyp) for key, dtyp in special_dtypes.items()}
        module_sizes = defaultdict(int)
    
        module_list = []
    
        if not buffers_only:
            module_list = named_module_tensors(model, recurse=True)
        else:
            #if hasattr(model, "named_buffers"):
>           module_list = model.named_buffers(recurse=True)
E           AttributeError: 'Parameter' object has no attribute 'named_buffers'

src/accelerate/utils/modeling.py:712: AttributeError
```

Reply:
I found the issue. This happens because you don't make sure that the `module` that you are passing in `infer_auto_device_map` is indeed a `module`. See comment below ! If the changes, it should be safe to remove the condition.